### PR TITLE
Adding a decorator to assert a user is not logged

### LINKED ISSUE #137: Need to assert a logged user can't access the register views

### PR DESCRIPTION
**Pull request by Micha546**
# Description
Adding a decorator to assert a user is not logged

Added the decorator in main.decorators.
Added the new decorator to the register views of Seeker
and Apartment.
Added tests to assert the decorator is working properly.

This PR is dependent on #147 - merged

used following refernces:
https://www.youtube.com/watch?v=eBsc65jTKvw&t=766s
https://docs.djangoproject.com/en/3.2/ref/contrib/auth/
https://www.datacamp.com/community/tutorials/decorators-python

## Screenshots
image

## Manual Tests
Did some tests to see that the decorator is working as intended

## Additional Information
Found a bug/issue with flake8, when i tried to import a fixture and use it flake8 was confused:
flake_error

We will probably have to redisign the fixtures in all the tests files, for now I added  `# noqa: F811` and `# noqa: F401` to
make flake8 ignore those false positives...
https://stackoverflow.com/questions/47196947/py-test-deal-with-both-pylint-and-flake8-when-importing-features-from-a-module

### Issues close by this PR
Fix #137

### CODE REVIEW

File: seekers/tests.py
Comment (Omeramsc) [inline seekers/tests.py]: You should not import fixtures but put them inside a conftest file. with that, they'll be automatically available for use across the scope they're in.

read more:
https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session

Don't ignore the linting issues, they're there for a reason.

**Review by Omeramsc — CHANGES_REQUESTED** [review]

File: seekers/tests.py
Comment (Micha546) [reply seekers/tests.py]: Yes, I saw that. but I think that moving all the fixtures to a new file and changing all the tests in the process is not in the scope of this PR. Can I open an issue for the redesigning of the fixtures and the tests and have this PR merged?

Comment (Micha546) [reply seekers/tests.py]: As I mentioned in the Additional Information part:
Found a bug/issue with flake8, when i tried to import a fixture and use it flake8 was confused,
We will probably have to redisign the fixtures in all the tests files, for now I added # noqa: F811 and # noqa: F401 to
make flake8 ignore those false positives...

Comment (Omeramsc) [reply seekers/tests.py]: No, what you should do in that case is first to open a PR that fixes this issue, and then rebase this PR on top of it.

Comment (Micha546) [reply seekers/tests.py]: Ok, I'll start working on it
👍

Comment (Micha546) [reply seekers/tests.py]: @Omeramsc I pushed a new PR that moves everything to conftest.py

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: this implicitly verifies that city_model is not None, so no need to explicitly check .
If city_model is None (or any object that doesn't have a city_name attribute) you'll get an AttributeError exception
`assert city_model is not None`
As an aside, if you're testing for not None you can leave that part out `assert city_model`

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: I'm unclear what you're checking here.
If you're trying to see if Django's get function works. There's no need to test that.

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: as above, no need to assert apartment_model is not none as it is implicit.
no need to use `is True`, `assert apartment_model.is_relevant` will do the same thing

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: same as above, what are you testing here?

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: Here you are testing a function that you wrote `get_apartment_by_id` so that is a good test.

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: You are passing in apartment_model as a parameter and it isn't used.

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: You are passing in user_model as a parameter and it isn't used.

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py]: Is the `valid_apartment_creation_form` not valid? Why is it supposed to raise an exception?

File: apartments/tests.py
Comment (simzacks) [inline apartments/tests.py:97]: why is user_model being passed in?

File: main/decorators.py
Comment (simzacks) [inline main/decorators.py:1]: Does this built in django decorator do what you're trying to do here?
https://docs.djangoproject.com/en/3.2/topics/auth/default/#the-login-required-decorator

File: main/tests.py
Comment (simzacks) [inline main/tests.py]: You're creating the same user in all the tests here. Won't that cause a problem?
You should probably use a setup function and create the user once

File: seekers/tests.py
Comment (simzacks) [inline seekers/tests.py]: why are you expecting this to fail?

File: seekers/tests.py
Comment (simzacks) [inline seekers/tests.py:55]: why pass in the user_model?

**Review by simzacks — CHANGES_REQUESTED** [review]

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: @simzacks Those changes are part of the #147 PR which this PR is dependent on.
I didn't write any new tests just moved fixtures to conftest.py

nevertheless I'll look at the changes you requested and change the code accordingly in #147

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: Ok, ill delete this test in #147

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: I'll delete this test in #147

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: Not my test, I only changed the fixture he is getting 👍

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: I think that without it the client wouldn't be able to login, but I'm not sure, it's not my test but ill try to run it without the apartment_model

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: same as above, ill try running it without it

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: With commit=True it's supposed to save the Apartment immediately to the data-base, because the form does not have a `owner` field I made it so the form throws an exception when you try to save it with commit=True, this test tests it

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py:97]: For the login, I'll try running it without it

File: main/decorators.py
Comment (Micha546) [reply main/decorators.py:1]: I wrote the opposite decorator, if you're logged in, you can't access the page

File: seekers/tests.py
Comment (Micha546) [reply seekers/tests.py:55]: For log-in ill try running it without it

File: seekers/tests.py
Comment (Micha546) [reply seekers/tests.py]: Yes, this form too has no `base-user` field so I made it throw an exception when trying to save with commit=True

File: main/tests.py
Comment (Micha546) [reply main/tests.py]: `make_user` was a fixture that did exactly what `get_user_model().objects.create_user` is doing, so I changed it

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: image

I've tried running it without the fixture and it failed.

File: seekers/tests.py
Comment (Micha546) [reply seekers/tests.py:55]: Does not work without the fixture

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py:97]: Does not work without the fixture

File: apartments/tests.py
Comment (Micha546) [reply apartments/tests.py]: Does not work without the fixture

**Comment by Micha546** [comment]
Just another heads-up, this PR is dependent on #147, so please merge #147 first

**Comment by simzacks** [comment]
Looks good. I'll wait until 147 is merged and this is rebased before approving

**Review by simzacks — APPROVED** [review]

**Review by OrAsaf — APPROVED** [review]

**Pull request merged**
